File: sliding_window/#209_minimum_size_subarray_sum/minSubArrayLen.js
/**
 * @param {number} target
 * @param {number[]} nums
 * @return {number}
 */
var minSubArrayLen = function (target, nums) {
  let start = 0;
  let end = 0;
  let total = 0;
  let minLength = Infinity;

  while (start < nums.length) {
    if (total < target && end < nums.length) {
      // 總和小於target就繼續遍歷, 但不能超過陣列總長
      total += nums[end];
      end++;
    } else if (total >= target) {
      // 總和已經超過target, 計算目前的長度, 並且移動start
      let currentLen = end - start;
      if (minLength > currentLen) {
        minLength = currentLen;
      }
      total -= nums[start];
      start++;
    } else if (end >= nums.length) {
      // 遍歷完陣列, 且總和沒超過target則結束
      break;
    }
  }

  if (minLength === Infinity) {
    console.log("cannot find minLen");
    return 0;
  }

  console.log(minLength);
  return minLength;
};

minSubArrayLen(7, [2, 3, 1, 2, 4, 3]); // 2
minSubArrayLen(4, [1, 4, 4]); // 1
minSubArrayLen(11, [1, 1, 1, 1]); // 0

File: sliding_window/#209_minimum_size_subarray_sum/minSubArrayLen.go
package main

import (
	"fmt"
	"math"
)

func main() {
	minSubArrayLen(7, []int{2, 3, 1, 2, 4, 3}) // 2
	minSubArrayLen(4, []int{1, 4, 4})          // 1
	minSubArrayLen(11, []int{1, 1, 1, 1})      // 0

}

func minSubArrayLen(target int, nums []int) int {
	start, end, total, minLen := 0, 0, 0, math.MaxInt32

	for start < len(nums) {
		if total < target && end < len(nums) {
			total += nums[end]
			end++
		} else if total >= target {
			currentLen := end - start
			if minLen > currentLen {
				minLen = currentLen
			}
			total -= nums[start]
			start++
		} else if end >= len(nums) {
			break
		}
	}

	if minLen == math.MaxInt32 {
		fmt.Println("cannot find minLen")
		return 0
	}

	fmt.Println(minLen)
	return minLen
}
